[css-scrollbars-1] Implement "scrollbar-color" support for iOS

nmoucht opened the pull request:
#### f0067bfaf1b378624705b77420639d727e5dad4a
<pre>
[css-scrollbars-1] Implement &quot;scrollbar-color&quot; support for iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=258461">https://bugs.webkit.org/show_bug.cgi?id=258461</a>
<a href="https://rdar.apple.com/111573662">rdar://111573662</a>

Reviewed by Tim Nguyen and Aditya Keerthi.

Adopt UIKit SPI for setting scrollbar color.

* LayoutTests/fast/scrolling/ios/scrollbar-color-body-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/scrollbar-color-body.html: Added.
* LayoutTests/fast/scrolling/ios/scrollbar-color-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/scrollbar-color.html: Added.
* LayoutTests/platform/ios-18/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setScrollingNodeScrollableAreaGeometry):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameScrollbarColor const):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _scrollbarState:processID:isVertical:]):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::mainFrameScrollbarColor const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeFrameScrollingNodeRemoteIOS.mm:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteIOS::scrollbarStateForOrientation const):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeOverflowScrollingNodeIOS.mm:
(WebKit::ScrollingTreeOverflowScrollingNodeIOS::scrollbarStateForOrientation const):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::commitStateAfterChildren):

Canonical link: <a href="https://commits.webkit.org/298826@main">https://commits.webkit.org/298826@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f2bc064a1f47c336e7613ac6b3acb17d385ea0d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116796 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36460 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27033 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122872 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67384 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4257320e-360f-419f-8d7a-d7b61821fd37) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118685 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37158 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45061 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88694 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43106 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bc30337b-762f-4fb6-86e2-1d85e0120b47) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119745 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29626 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104773 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69164 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28689 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22878 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66537 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108910 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99005 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23034 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126006 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115324 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43695 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32817 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97362 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44059 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100975 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97160 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24734 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42486 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20426 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39703 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43581 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49177 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144022 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43048 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/37082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46387 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44753 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7af98b0953257de49794c4a9749a57a37f92e163. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a1dc294af4a935b7d10262739d942a02dd8681ff. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

pxlcoder commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1762,6 +1762,11 @@
 #endif
 
+#if !defined(HAVE_UIKIT_SCROLLBAR_COLOR_SPI) \
+    && ((PLATFORM(IOS_FAMILY) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 260000))
```
The combination of `PLATFORM(IOS_FAMILY)` and `__IPHONE_OS_VERSION_MIN_REQUIRED` is strange.

If you want it on all iOS family, list out the individual platforms like below. If not, then use `(PLATFORM(IOS) || PLATFORM(MACCATALYST)` instead of `PLATFORM(IOS_FAMILY)`.

pxlcoder commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeOverflowScrollingNodeIOS.mm:
```diff
@@ -94,9 +99,15 @@
 String ScrollingTreeOverflowScrollingNodeIOS::scrollbarStateForOrientation(ScrollbarOrientation orientation) const
 {
-    if (auto* scrollView = this->scrollView()) {
+    if (UIScrollView *scrollView = this->scrollView()) {
```
```suggestion
    if (RetainPtr scrollView = this->scrollView()) {
```

nmoucht replied:
Fixed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/243ff43a5eb917c59519764be8d0949415f1fa9f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

nmoucht commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1000,6 +1000,16 @@ - (void)_updateScrollViewForTransaction:(const WebKit::RemoteLayerTreeTransactio
+    auto scrollbarColor = _page->scrollingCoordinatorProxy()->mainFrameScrollbarColor();
+    if (scrollbarColor) {
+        [_scrollView _setHorizontalScrollIndicatorColor:cocoaColor(scrollbarColor->thumbColor).get()];
+        [_scrollView _setVerticalScrollIndicatorColor:cocoaColor(scrollbarColor->thumbColor).get()];
```
Fixed.

nmoucht commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeFrameScrollingNodeRemoteIOS.mm:
```diff
@@ -163,9 +170,15 @@
 String ScrollingTreeFrameScrollingNodeRemoteIOS::scrollbarStateForOrientation(ScrollbarOrientation orientation) const
 {
-    if (auto* scrollView = this->scrollView()) {
+    if (UIScrollView *scrollView = this->scrollView()) {
```
Fixed.

nmoucht commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
```diff
@@ -406,6 +407,25 @@ - (void)scrollView:(WKBaseScrollView *)scrollView handleScrollUpdate:(WKBEScroll
+
+        if (scrollbarColor) {
+            [scrollView _setHorizontalScrollIndicatorColor:cocoaColor(scrollbarColor->thumbColor).get()];
+            [scrollView _setVerticalScrollIndicatorColor:cocoaColor(scrollbarColor->thumbColor).get()];
```
Fixed.

nmoucht commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1762,6 +1762,11 @@
 #endif
 
+#if !defined(HAVE_UIKIT_SCROLLBAR_COLOR_SPI) \
+    && (((PLATFORM(IOS) || PLATFORM(MACCATALYST)) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 260000))
```
Fixed.

pxlcoder commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
```diff
@@ -89,6 +89,7 @@ struct ScrollableAreaParameters {
     NativeScrollbarVisibility verticalNativeScrollbarVisibility { NativeScrollbarVisibility::Visible };
 
     ScrollbarWidth scrollbarWidthStyle { ScrollbarWidth::Auto };
+    std::optional<ScrollbarColor> scrollbarColorStyle { };
```
```suggestion
    std::optional<ScrollbarColor> scrollbarColorStyle;
```

`std::optional` is default initialized to nullopt.

nmoucht replied:
Fixed.

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
```diff
@@ -330,6 +330,9 @@ - (NSString *)_scrollbarState:(unsigned long long)rawScrollingNodeID processID:(
         {
             TextStream::GroupScope scope(ts);
             ts << ([_scrollView showsHorizontalScrollIndicator] ? ""_s : "none"_s);
+#if HAVE(UIKIT_SCROLLBAR_COLOR_SPI)
+            ts << ([_scrollView _verticalScrollIndicatorColor] ? WebCore::colorFromCocoaColor([_scrollView _verticalScrollIndicatorColor]).debugDescription() : ""_s);
+#endif
```
I think we're missing the equivalent logging change here? To include horizontal when necessary.

nmoucht replied:
Fixed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ee347a59d1f0968f688db684fecda0e5da3c94a7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/257346846a62f5ea3a99585936a3c1957a07fd3c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b571a4266aa01ecc00064ac02fc4d300a9d97996. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3bb1bfaff51ae7b0ed611865f2cc53c15f375298. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ac1feb039708de9210f5dab34774a1ac5852d59a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a196c300aba2d5a60dcb151c83a7da44b9e2415d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/116fbe3ce4530cb7645007e26bd8c4cff04ab2fc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6de8effb6af3ebcc06f1ce8aaf14794a1b0ae53f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1e5261fb1a7ae78806b68890d3bf48b0f90f0adf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b932879972c4a671bea6cfa536bfdf257ffce16c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4dc77fca50279f24ba49da5e4e37a95f78e6c772. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

nt1m commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeOverflowScrollingNodeIOS.mm:
```diff
@@ -94,9 +99,18 @@
-        return showsScrollbar ? ""_s : "none"_s;
+        ts << (showsScrollbar ? ""_s : "none"_s);
+#if HAVE(UIKIT_SCROLLBAR_COLOR_SPI)
+        if (orientation == ScrollbarOrientation::Horizontal)
+            ts << ([scrollView _horizontalScrollIndicatorColor] ? colorFromCocoaColor([scrollView _horizontalScrollIndicatorColor]).debugDescription() :  ""_s);
+        else
+            ts << ([scrollView _verticalScrollIndicatorColor] ? colorFromCocoaColor([scrollView _verticalScrollIndicatorColor]).debugDescription() :  ""_s);
```
```suggestion
        if (orientation == ScrollbarOrientation::Horizontal)
            ts << ([scrollView _horizontalScrollIndicatorColor] ? WebCore::colorFromCocoaColor([scrollView _horizontalScrollIndicatorColor]).debugDescription() :  ""_s);
        else
            ts << ([scrollView _verticalScrollIndicatorColor] ? WebCore::colorFromCocoaColor([scrollView _verticalScrollIndicatorColor]).debugDescription() :  ""_s);
```
to fix internal builds

nmoucht replied:
Fixed.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5bef722aca162662a8fc61f1b765c2ccb4e7bd8a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49248

nt1m commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -119,7 +119,11 @@ ScrollingStateScrollingNode::ScrollingStateScrollingNode(const ScrollingStateScr
+#if PLATFORM(MAC) || PLATFORM(IOS_FAMILY)
     , m_scrollbarColor(stateNode.scrollbarColor())
+#endif
+#if PLATFORM(MAC)
```
Could you address this comment?

nt1m commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -2899,8 +2899,16 @@ header: <WebCore/ScrollTypes.h>
     ReplacedByCustomScrollbar
 };
 
+#if !PLATFORM(COCOA)
+header: <WebCore/ScrollTypes.h>
+[Nested, CustomHeader] struct WebCore::ScrollbarColor {
+    WebCore::Color thumbColor;
+    WebCore::Color trackColor;
+};
+#endif
+
```
Does it make sense to remove the equivalent in RemoteScrollingCoordinatorTransaction.serialization.in and just make this cross-platform?

nmoucht committed:
[css-scrollbars-1] Implement "scrollbar-color" support for iOS
https://bugs.webkit.org/show_bug.cgi?id=258461
rdar://111573662

Reviewed by Tim Nguyen and Aditya Keerthi.

Adopt UIKit SPI for setting scrollbar color.

* LayoutTests/fast/scrolling/ios/scrollbar-color-body-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/scrollbar-color-body.html: Added.
* LayoutTests/fast/scrolling/ios/scrollbar-color-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/scrollbar-color.html: Added.
* LayoutTests/platform/ios-18/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setScrollingNodeScrollableAreaGeometry):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameScrollbarColor const):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _scrollbarState:processID:isVertical:]):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::mainFrameScrollbarColor const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeFrameScrollingNodeRemoteIOS.mm:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteIOS::scrollbarStateForOrientation const):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeOverflowScrollingNodeIOS.mm:
(WebKit::ScrollingTreeOverflowScrollingNodeIOS::scrollbarStateForOrientation const):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::commitStateAfterChildren):

Canonical link: https://commits.webkit.org/298826@main
